prov/efa/src/rxr/rxr_pkt_type_data.c

```diff
@@ -36,207 +36,6 @@
-	/* Constructed iov's total size */
-	uint64_t payload_size = 0;
-	/* pkt_entry offset to write data into */
-	uint64_t pkt_used = 0;
-	uint64_t orig_iov_index;
-	uint64_t orig_iov_offset;
-	/* Remaining size that can fit in the constructed iov */
-	uint64_t remaining_len = MIN(tx_entry->window,
-				     ep->max_data_payload_size);
-	/* The constructed iov's index */
-	size_t i = 0;
-	size_t len = 0;
-
-	size_t j;
-	ssize_t ret;
-
-	orig_iov_index = tx_entry->iov_index;
-	orig_iov_offset = tx_entry->iov_offset;
-
-	data_pkt = (struct rxr_data_pkt *)pkt_entry->pkt;
-	/* Assign packet header in constructed iov */
-	iov[i].iov_base = rxr_pkt_start(pkt_entry);
-	iov[i].iov_len = sizeof(struct rxr_data_hdr);
-	desc[i] = fi_mr_desc(pkt_entry->mr);
-	i++;
-
-	/*
-	 * Loops until payload size is at max, all user iovs are sent, the
-	 * constructed iov count is greater than the core iov limit, or the tx
-	 * entry window is exhausted.  Each iteration fills one entry of the
-	 * iov to be sent.
-	 */
-	while (tx_entry->iov_index < tx_entry->iov_count &&
-	       remaining_len > 0 && i < ep->core_iov_limit) {
-		if (tx_entry->desc[tx_entry->iov_index]) {
-			iov[i].iov_base =
-				(char *)tx_iov[tx_entry->iov_index].iov_base +
-				tx_entry->iov_offset;
-			desc[i] = tx_entry->desc[tx_entry->iov_index];
-
-			len = tx_iov[tx_entry->iov_index].iov_len
-			      - tx_entry->iov_offset;
-			if (len > remaining_len) {
-				len = remaining_len;
-				tx_entry->iov_offset += len;
-			} else {
-				tx_entry->iov_index++;
-				tx_entry->iov_offset = 0;
-			}
-			iov[i].iov_len = len;
-		} else {
-			/* It should be noted for cuda buffer, caller will always
-			 * provide desc, and will not enter this branch.
-			 *
-			 * Copies any consecutive small iov's, returning size
-			 * written while updating iov index and offset
-			 */
-
-			len = rxr_copy_from_iov((char *)data_pkt->data +
-						 pkt_used,
-						 remaining_len,
-						 tx_entry);
-
-			iov[i].iov_base = (char *)data_pkt->data + pkt_used;
-			iov[i].iov_len = len;
-			desc[i] = fi_mr_desc(pkt_entry->mr);
-			pkt_used += len;
-		}
-		payload_size += len;
-		remaining_len -= len;
-		i++;
-	}
-
-	pkt_entry->send = ofi_buf_alloc(ep->pkt_sendv_pool);
-	if (!pkt_entry->send) {
-		FI_WARN(&rxr_prov, FI_LOG_EP_DATA,
-			"Unable to allocate rxr_pkt_sendv from pkt_sendv_pool\n");
-		return -FI_EAGAIN;
-	}
-
-	for (j = 0; j < i; j++) {
-		memcpy(&pkt_entry->send->iov[j], &iov[j], sizeof(struct iovec));
-		pkt_entry->send->desc[j] = desc[j];
-	}
-	pkt_entry->send->iov_count = i;
-
-	data_pkt->hdr.seg_length = (uint16_t)payload_size;
-	pkt_entry->pkt_size = payload_size + sizeof(struct rxr_data_hdr);
-	pkt_entry->x_entry = tx_entry;
-	pkt_entry->addr = tx_entry->addr;
-
-	FI_DBG(&rxr_prov, FI_LOG_EP_DATA,
-	       "Sending an iov count, %zu with payload size: %lu.\n",
-	       i, payload_size);
-	ret = rxr_pkt_entry_send(ep, pkt_entry, tx_entry->send_flags);
-	if (OFI_UNLIKELY(ret)) {
```

Comment:
Can you elaborate on this?

I think we only resend when `rxr_pkt_entry_send` return -FI_EAGAIN. The resend was handled by the progress engine `rxr_ep_progress_internal`?

Reply:
Where this comment is there is some code that resets `bytes_sent` `tx_entry->iov_offset` on failure, so that segment can be resent later. It looks like there is a check at the beginning of `rxr_pkt_post_data` for no TX packets available, and all other errors are fatal.